main.py

```diff
@@ -180,6 +180,258 @@ def canonical_idx4_reverse(ijkl):
+    +-------+-------------------+---------------+-----------------+-------------------------------+------------------------------+-----------------------------+
+    |       | i<j<k<l (1,2,3,4) |   <    <   <  |  <   <   <   <  |                               | 1<->3 x 2<->4                |                             |
+    |   G   | i<k<j<l (1,3,2,4) |   <    <   <  |  <   >   <   <  |                               | 1<->2 x 3<->4                |                             |
+    |       | j<i<k<l (2,1,3,4) |   <    <   <  |  >   <   <   <  |                               | 1<->4 x 2<->3                |                             |
+    +-------+-------------------+---------------+-----------------+-------------------------------+------------------------------+-----------------------------+
+    """
+    assert (i, j, k, l) == canonical_idx4_reverse(compound_idx4(i, j, k, l))
+    if i == l:
+        return "A"
+    elif (i == k) and (j == l):
+        return "B"
+    elif (i == k) or (j == l):
+        if j == k:
+            return "D"
+        else:
+            return "C"
+    elif j == k:
+        return "E"
+    elif (i == j) and (k == l):
+        return "F"
+    elif (i == j) or (k == l):
+        return "E"
+    else:
+        return "G"
+
+
+def test_pair_idx(dadb, idx, category):
+    foo = {
+        "A": test_pair_idx_A,
+        "B": test_pair_idx_B,
+        "C": test_pair_idx_C,
+        "D": test_pair_idx_D,
+        "E": test_pair_idx_E,
+        "F": test_pair_idx_F,
+        "G": test_pair_idx_G,
+    }[category]
+    return foo(dadb, idx)
+
+
+def test_pair_idx_A(dadb, idx):
+    da, db = dadb
+    assert integral_category(*idx) == "A"
+    i, _, _, _ = idx
+    assert (i, i, i, i) == idx
+    assert da == db
+    assert (i in da.alpha) and (i in da.beta)
+    return
+
+
+def test_pair_idx_B(dadb, idx):
+    da, db = dadb
+    assert integral_category(*idx) == "B"
+    i, j, _, _ = idx
+    assert (i, j, i, j) == idx
+    assert da == db
+    assert (i in da.alpha + da.beta) and (j in da.alpha + da.beta)
+    return
+
+
+def test_pair_idx_C(dadb, idx):
+    da, db = dadb
+    assert integral_category(*idx) == "C"
+    i, j, k, l = idx
+    exc = Excitation.exc_degree(da, db)
+    if exc == (1, 0):
+        dsa = da.alpha
+        dsb = db.alpha
+    elif exc == (0, 1):
+        dsa = da.beta
+        dsb = db.beta
+    else:
+        raise
+    _, h, p = PhaseIdx.single_exc(dsa, dsb)
+    if j == l:
+        assert sorted((h, p)) == sorted((i, k))
+        assert (j in da.alpha + da.beta) and (j in db.alpha + db.beta)
+    elif i == k:
+        assert sorted((h, p)) == sorted((j, l))
+        assert (i in da.alpha + da.beta) and (i in db.alpha + db.beta)
+    else:
+        raise
+    return
+
+
+def test_pair_idx_D(dadb, idx):
+    da, db = dadb
+    assert integral_category(*idx) == "D"
+    i, j, k, l = idx
+    exc = Excitation.exc_degree(da, db)
+    if exc == (1, 0):
```

Comment:
I can try and do this if Kevin hasn't already started it. You mean getting the spin determinants involved in some excitation?

Reply:
I haven't started yet, but I'm not sure exactly what @TApplencourt is describing

Reply:
We have some code duplications of  a pattern that look like roughly like this:
```
        if exc == SOMETHING:
            return alpha/beta part of some determinant
```
If we can create a function, it will make the code terser.

Reply:
Do the s, t in dsa, dta indicate anything special @kgasperich ? Or just the notation you chose

Reply:
I wanted to have different labels for different spins, so I chose `s` and `t`, so in some cases `(s, t) == (a, b)`, and sometimes `(s, t) == (b, a)` (and in cases where we are only considering one of the spins I just used `s`)

Reply:
> We have some code duplications of a pattern that look like roughly like this:
> 
> ```
>         if exc == SOMETHING:
>             return alpha/beta part of some determinant
> ```
> 
> If we can create a function, it will make the code terser.

We have several cases where we will want different things, so I'm not sure if that will make it any more readable (it might just obfuscate it more).
I can simplify the assignment with something like `(dsa, dta), (dsb, dtb) = da, db`